### ISSUE TITLE
bug(decode): Qwen3-dense emits a single token id 0 in ~50% of runs under Metal shader validation — silently, exit 0

### ISSUE THREAD

**Issue by Pushkinist** [Pushkinist/rMLX#346]
## Symptom

Under Metal shader validation, `Ternary-Bonsai-8B-mlx-2bit` (Qwen3 dense) emits a
**single token, id 0 (`"!"`)**, then stops. `decode_tps` ~0.064, `WARN … store
sequence did not advance`, **process exits 0**. A healthy run at the same cell
produces 63–64 tokens.

It is non-deterministic and roughly coin-flip:

| cell | validated | unvalidated |
|---|---|---|
| Bonsai `--kv-quant k8v8` | 6 / 12 degenerate | 0 / 13 |
| Bonsai `--kv-quant none` | 2 / 4 degenerate | — |
| gemma-4-e2b (all cells) | 0 / 9 | 0 / — |

Fisher exact on the pooled validated-vs-unvalidated split: **p ≈ 0.005**.

Reproduce (4096-token prompt, 64 generated, scratch `RMLX_HOME`, `--metrics off`):

```
bash scripts/run_gpu_tests.sh --help   # for the pinned MTL_SHADER_VALIDATION_* set
```

then run a normal generate under that same pinned environment. Repeat ~8×; roughly
half degenerate.

## What has been ruled out

- **Not an out-of-bounds device store.** `MTL_SHADER_VALIDATION_FAIL_MODE` is the
  discriminator and it comes out **inverted**: clean 3/3 under `zerofill` (which
  *drops* invalid writes) and reproduced 1/3 under `allow` (which *permits* them).
  A zerofilled store as the cause predicts the opposite assignment.
- **No diagnostic anywhere.** No `Invalid …` on stderr under the pinned environment
  with `MTL_SHADER_VALIDATION_REPORT_TO_STDERR=1`, and `log show --last 20m
  --predicate 'process == "rmlx"'` returns 0 matching lines. Validation itself never
  complains.
- **Not a prefill-duration watchdog.** gemma-4-e2b with a 17.7–19.3 s *validated*
  prefill (32k prompt) is 3/3 clean, 9/9 across all validated cells.
- **Not codec-specific.** Reproduces at both `k8v8` and `none`.
- **Not reachable from the unit suites.** The `rmlx-kv-quant` GPU suite is
  mode-invariant: 234 passed / same 4 known-red / 0 invalid reports, in three reps
  of each mode. The unit shapes never hit it.
- **Token output is otherwise bit-identical under validation** where it does not
  degenerate — verified by hashing `token_id` trace events for
  gemma-4-e2b/`k8v8`, gemma-4-e2b/`iso3_sym`, Bonsai/`iso3_sym`.

## Why this is worth resolving rather than documenting away

Two readings remain, and they have opposite consequences:

1. **A latent engine defect that instrumentation perturbs into visibility.** Shader
   validation changes dispatch timing substantially (+18% on the GPU test suite,
   5.5× on real decode), which is exactly the kind of perturbation that exposes a
   race or an ordering assumption in the decode path. `store sequence did not
   advance` is a KV-store symptom, and this repo has a documented history of silent
   KV corruption producing wrong output with no error.
2. **An instrumentation defect** — the validation layer mis-handling something in
   the Qwen3-dense path on this host, with no consequence for uninstrumented runs.

Under reading 1 there is a real bug reachable without validation, at some lower
rate, on a code path that produces wrong output and exits 0. That is the repo's
highest-severity class. Until it is discriminated, the honest statement is the one
now in `docs/TESTING.md`: never draw a conclusion about *model output* from a
validated run.

## Suggested next probes

- Does it reproduce with validation off but the same dispatch perturbation applied
  some other way (`MTL_DEBUG_LAYER=1`, deliberate host-side delay per decode step)?
  That separates "timing" from "validation".
- What does `store sequence did not advance` mean at the moment it fires — capture
  the KV store state and the step index under `--log verbose`.
- Does the rate change with generation length, prompt length, or thread count?
- Does another Qwen3-dense snapshot reproduce it, or only this one?

Found while validating #328. Not caused by that work: the same environment is what
made it observable, and the branch ships only a docs warning for it.

**Comment by Pushkinist** [comment]
## Diagnosis — two layers. The first is certain and fixable now; the second is not localized.

108 validated runs, four separately-built binaries, on a host whose quiescence was measured rather than assumed.

## Layer 1 — why it is silent, exits 0, and becomes a recorded number

`rmlx baseline` passes `&[]` for `eos_ids` ("no EOS-stop, force full max_tokens steps"). Given that, exactly **one** path in `qwen3::generate_greedy` returns `Ok` with one token *before* reaching the KV-byte store — which is what makes the `store sequence did not advance` warning fire:

```rust
// crates/rmlx-models/src/qwen3.rs:2132
if nan_count > 0 {
    return Ok(steps);
}
```

The prefill token has already been pushed and handed to `step_fn` twelve lines above, so the caller gets one token; the store at `:2225-2228` is never reached; the function returns success. **The condition is detected and then discarded — the site logs nothing at any level.**

So the prefill logits contained NaN, `argmax` over an all-NaN vector returned index 0, and token id 0 is `"!"`.

Three measurements confirm the reading:

1. The reported `tps=0.064` is arithmetically pinned to `n_generated == 1`. `compute_phase_timing` only falls back to `overall_tps = n/total_s` when `n_generated < 2`. A validated 3770-token prefill on this host runs 246–266 tok/s → 14.2–15.3 s → `1/15.3 = 0.065`.
2. A captured degenerate run: `ttft_ms=15299 decode_tps=0.065 prefill_tps=246.4`, `n_generated=1`, `decoded="!"`, one WARN, **zero ERROR, exit 0**.
3. The warn discriminates correctly — a forced `--max-tokens 1` run *does* reach the KV store (`warn=0`), so `warn=1` is specific to a pre-store early return.

**Seven sites share the shape**: `return Ok(steps)` in qwen3, gemma4, gemma3, qwen2, bitnet, laguna; a `break` in laguna's decode loop.

### Three coverage holes, each a check that cannot fail here

- **`make check-no-decode-swallow` is structurally blind.** RULE 1 anchors on `error = %e`, "the structural marker shared by every failure-log site in the arch layer". A swallow that logs *nothing* has no anchor — while producing exactly the shape the gate's own header names: "a fabricated zero, emitted as a measurement, with exit 0".
- **`classify_smoke` has a `BrokenNan` verdict and no caller on any real generation** — only `rmlx info --probe` and load-time `--require-smoke-probe`.
- **Two architectures never compute `nan_count` at all.** `qwen3_5_moe` and `qwen3_vl_moe` hard-code `nan_count: 0`, so on Qwen3.6-35B-A3B — a required test-target family — a NaN prefill produces a *full-length* run of garbage with no guard and no verdict.

`rmlx bench` already refuses this shape. `rmlx baseline` downgrades the identical verdict to `warn!` + `0` and prints a summary; with `--record` that writes a permanent row into the append-only `runs.db`. The server returns **HTTP 200** carrying the single token `"!"`.

## The trigger is host regime — not the codec, not the commit

3 degenerate events in 108 runs. **All three consecutive, all three inside one window where `prefill_tps` was depressed ~7%.**

| campaign | binary | prefill_tps | n | degen |
|---|---|---|---|---|
| HEAD, various | `f87f149e` | 260–266 | 32 | 0 |
| **OLD, info** | **`1e6097f2`** | **246.4–249.3** | 12 | **3** (runs 4,5,6) |
| OLD, debug | `1e6097f2` | 265.7–266.3 | 12 | 0 |
| `ff98544b` / `4530929e` | | 265.4–266.4 | 24 | 0 |
| 48 CPU spinners, both arms | | 235–244 | 20 | 0 |
| OLD, 16k prompt | `1e6097f2` | 164 | 3 | 0 |

**0/60 at ~266 tok/s across four binaries; 3/10 at ~247.** The three failures were the three slowest prefills in that campaign; runs 11–12 jumped to 262.8/266.1 and went clean.

Ruled out with numbers: **not `--log debug`** (same binary, 0/12 at debug vs 3/10 at info — the extra per-chunk events perturb timing enough to suppress it); **not CPU contention** (48 spinners depressed prefill *below* the reproducing window and halved decode — a different signature; 0/20); **not prompt length** (16k prompt, 0/3).

### I am retracting my own A/B

I built the issue-era tree and got 3/12 vs 0/32 on HEAD, Fisher p = 0.0136, then bisected toward the device-mask commit. Both bisect arms came back 0/12 — **and every one of those arms ran at 265–266 while the reproducing campaign ran at 247. HEAD has never been tested in the reproducing regime.** The p-value is confounded by host state and I withdraw it. Session totals are 3/42 vs 0/42, p ≈ 0.24, and the three events are not independent draws anyway. **I cannot say whether HEAD is fixed.**

Binary distinctness was verified both times (`sparse_v_weighted_sum` 7 vs 0 occurrences; the `layers/mask.rs` panic-location string present in one arm and absent in the other).

## The `inv=0` negative is real, not a vacuous scan

Across all 108 runs including the 3 degenerate ones: validation banner present, **zero** matches for an invalid-access report. The detector was mutation-checked by running the OOB canary under **both** `FAIL_MODE=zerofill` and `FAIL_MODE=allow` — banner 1, diagnostic 1, in both. So an out-of-bounds *access* would have been reported and was not.

Note the issue's `FAIL_MODE` discriminator was n=3 per arm, Fisher p = 1.0 — no power at any effect size. And its direction was read backwards: `allow` reproducing / `zerofill` clean is what a genuine OOB access predicts; it is inverted only for "validation drops a write the engine needs", a different hypothesis.

## Layer 2 — where the NaN comes from

Not localized. What the evidence constrains: Qwen3-dense **prefill**, not decode; timing-sensitive at finer granularity than one `debug!` per chunk; gated on a host regime depressing prefill ~7% while leaving decode untouched; and **no** shader-validation diagnostic from a detector proven live under both fail modes.

That combination fits an **in-bounds read of stale or never-written device memory** better than an out-of-bounds access — validation catches the latter and not the former, non-determinism follows from what the MLX buffer cache last left there, and allocation-state sensitivity follows from which recycled buffer you get. Natural candidate: the KV ring, sized `next_pow2_seq(needed_seq)` — at a 3770-token prompt the ring is 4096, leaving ~326 slots never written. One NaN in a softmax numerator makes the whole row NaN, which propagates through the residual stream to every logit.

**This is the hypothesis to test, not a conclusion.**

## Proposed fix

**Part A — a NaN prefill must not return success.** At all six `return Ok(steps)` sites: `error!` with `nan_count`, `max_abs_logit`, `prompt_len`, and propagate. Two decisions for whoever implements it: `Error::Model` vs `Error::Other` (the former is used for the deterministic empty-prompt case *specifically so the retry envelope will not replay it* — this failure is non-deterministic, so Migratable may be correct; check `docs/SERVER.md`), and whether laguna's mid-decode `break` should truncate-with-error or error out.

The `error = %e` field shape matters independently: it gives `check-no-decode-swallow` RULE 1 an anchor at these sites for the first time.

**Part B — close the detector holes.** Compute `nan_count` in `qwen3_5_moe` and `qwen3_vl_moe` (without this, Part A does nothing on Qwen3.6); add a `check-no-decode-swallow` rule for *detection without logging*; make `rmlx baseline` refuse `KvBytesVerdict::Unreported` as `rmlx bench` already does.

**Part C — Layer 2 is not fixed by any of the above.** Part A converts a wrong answer into a loud failure; it does not stop the NaN. **Do not close this issue on Part A alone.**

### The Layer-2 experiment nobody has run

Hold the ring at 4096 and vary the never-written tail — the one test that separates "stale-memory read" from everything else:

| arm | prompt | ring | uninitialized tail |
|---|---|---|---|
| T-large | 2100 | 4096 | ~1930 slots |
| T-small | 4000 | 4096 | ~30 slots |

Bonsai-8B, `k8v8`, `--max-ctx 4096`, `--log info`, `FAIL_MODE=allow`, **ABBA-interleaved** (never blocked — host drift already invalidated one A/B here), 30 runs per arm.

**Pass criterion, stated before the run:** confirmed if T-large ≥5/30 while T-small ≤1/30 (Fisher p ≤ 0.05). **Falsified** if the arms are within 2 events. If *both* are 0/30, the trigger was not present that session — report the `prefill_tps` band and conclude nothing, exactly as happened to me.

Record `prefill_tps` on every run and do not pool across a >5% shift. A campaign that never enters the 246–249 band has not sampled the failure regime, and its zeros are uninformative.

**Comment by Pushkinist** [comment]
Part C investigated. **Not fixed** — it is a research question, not a defect with a known site — but the suspect set is now much smaller, and one experiment this issue proposes should not be run.

## The Layer-2 hypothesis is false by construction — do not spend the GPU time

This issue proposes that the NaN comes from an in-bounds read of the never-written KV-ring
tail (`next_pow2_seq`, 3770 → ring 4096, ~326 unwritten slots), and specifies a 60-run
T-large / T-small campaign to test it. Two facts kill it:

- Both the lazy alloc (`kvcache/update.rs:2634-2635`) and the grow path (`:2380-2381`) use
  `zeros()` → `mlx_zeros`. The tail is 0.0, not recycled memory.
- `update.rs:2655-2659` returns `slice(buf, 0..offset)`. **The tail is never in an SDPA input
  at any prompt length.**

The two arms of the proposed experiment differ only in the length of a slice that is then
sliced off.

## No rMLX Metal kernel runs before that logit row on this architecture

While `in_prefill` is set, `KvCache::update` returns at `update.rs:2147-2149` into
`update_prefill_raw`, which is MLX ops only. The codec dispatch sits behind that branch
(`:2162-2168`), fused-QK is gated on `q_seq == 1`, Qwen3-dense has no rotating layers, and
neither `rmlx-models` `.metal` kernel (`gated_delta_step`, `paroquant_rotate`) is on this
arch/checkpoint. Every kernel rMLX dispatches in such a request comes from `exit_prefill`,
which runs *after* the logit graph is built and takes the ring as a read-only input.

That is why the earlier campaign's zero invalid-access reports were never going to fire —
now for a structural reason rather than an unexplained negative.

## Two premises in the issue text are weaker than they look

- **"Reproduces at both `k8v8` and `none`" is vacuous as a codec discriminator.** Those cells
  execute the *same* prefill appends over the same bf16 buffer. (`none` also force-K8V8s
  head/tail layers, but only post-prefill.) The observation is consistent with a shared-path
  fault and carries nothing beyond that.
- **gemma-4-e2b's 0/9 is not the counter-evidence it appears to be.** `enter_prefill` returns
  early on a rotating cache (`update.rs:2680`), so gemma-4's SWA layers never touch the shared
  append — only its `full_attention` layers do, where all 36 Qwen3-dense layers do. Equal run
  counts are not equal draws.

## What is left points at MLX, not at us

MLX allocates with `MTLHazardTrackingModeUntracked`, and its Metal command encoder models
read-after-write hazards only, not write-after-read. Upstream `ml-explore/mlx#3630` is a real
intermittent-wrong-value bug of exactly this shape, and `ml-explore/mlx#3461` covers buffer
lifetime under untracked mode — filed from this same host and hardware. The pin here is MLX
0.31.2.

One coupling was suspected and ruled out: `exit_prefill` calls a blocking `eval()`
(`update.rs:2792`), so its kernels are not concurrent with the pending logits graph.

The `nan_count` detector itself is correct (`qwen3.rs:1780-1795`, bf16 via `<< 16`), so the
recorded events were genuine NaN rather than a false positive.

## Landed on `fix/346-part-c`

Two CPU regression tests pinning the exclusions above, and a `docs/TESTING.md` correction
replacing the open "stale or never-written device memory" speculation. No runtime behaviour
changes.

Both tests are mutation-checked. Worth recording one detail: the tail check counts elements
`!= 0.0` rather than folding a max-of-abs, because `f32::max` propagates the other operand
past a NaN — the first version of that fold reported a clean tail *under a NaN mutation*.

## What actually blocks Part C

**No degenerate run has yet been captured with Part A's instrumentation** — the 108-run
campaign predates that merge. That single measurement is the blocker, and no further code is
needed to obtain it. `nan_count == vocab` would mean the whole hidden row went NaN
(attention/norm); a small `nan_count` with a finite `max_abs_logit` would mean isolated cells
out of `lm_head`, which is a different fault entirely.

**A bounded negative campaign cannot prove a fix here.** The pooled rate is 3/108 = 2.8%, so
105 independent runs would be needed for 95% confidence of seeing one — but the events are not
independent: all three came from the same 10-run in-band window, and the band is not
controllable (CPU spinners pushed *below* it; `--log debug` suppressed it). Budget any campaign
as evidence capture, not as a gate.
